File: tuan2/PhanC/bai49.jl
using Printf

struct Mau
    id::Int
    x::Vector{Float64}
    nhan::String
end

kc_euclid(u::Vector{Float64}, v::Vector{Float64}) = sqrt(sum((u .- v) .^ 2))
kc_manhattan(u::Vector{Float64}, v::Vector{Float64}) = sum(abs.(u .- v))

function danh_gia_loocv(D::Vector{Mau}, ten_do_do::String, kc_func::Function)
    n = length(D)
    dung = 0
    ds_sai = String[]

    for i in 1:n
        kc_min = Inf
        du_doan = ""

        for j in 1:n
            if i == j continue end # Bỏ qua chính mẫu đang xét
            dist = kc_func(D[i].x, D[j].x)
            if dist < kc_min
                kc_min = dist
                du_doan = D[j].nhan
            end
        end

        if du_doan == D[i].nhan
            dung += 1
        else
            push!(ds_sai, "Mau $(D[i].id) (that: $(D[i].nhan), du doan: $du_doan)")
        end
    end

    ti_le = dung / n * 100.0
    @printf("%s: %d/%d = %.2f%%\n", ten_do_do, dung, n, ti_le)
    if isempty(ds_sai)
        println("  -> Khong co mau nao bi sai.")
    else
        println("  -> Cac mau bi sai:")
        for s in ds_sai
            println("     $s")
        end
    end
end

function main()
    if !isfile("hoa30.txt")
        println(stderr, "Khong the mo tep hoa30.txt")
        return
    end

    lines = filter(!isempty, strip.(readlines("hoa30.txt")))
    n, d = parse.(Int, split(lines[1]))

    D = Vector{Mau}(undef, n)
    for i in 1:n
        parts = split(lines[i + 1])
        x = parse.(Float64, parts[1:d])
        nhan = String(parts[d + 1])
        D[i] = Mau(i, x, nhan)
    end

    println("=== KET QUA DANH GIA LOOCV TREN HOA30.TXT ===")
    danh_gia_loocv(D, "Euclid", kc_euclid)
    danh_gia_loocv(D, "Manhattan", kc_manhattan)
end

main()

File: tuan2/PhanC/bai49.py
import sys
import math

def kc_euclid(u, v):
    return math.sqrt(sum((u[k] - v[k]) ** 2 for k in range(len(u))))

def kc_manhattan(u, v):
    return sum(abs(u[k] - v[k]) for k in range(len(u)))

def danh_gia_loocv(D, ten_do_do, kc_func):
    n = len(D)
    dung = 0
    ds_sai = []

    for i in range(n):
        kc_min = float("inf")
        du_doan = None

        for j in range(n):
            if i == j:
                continue # Bỏ qua chính mẫu đang xét
            dist = kc_func(D[i][1], D[j][1])
            if dist < kc_min:
                kc_min = dist
                du_doan = D[j][2]

        if du_doan == D[i][2]:
            dung += 1
        else:
            ds_sai.append(f"Mau {D[i][0]} (that: {D[i][2]}, du doan: {du_doan})")

    ti_le = dung / n * 100.0
    print(f"{ten_do_do}: {dung}/{n} = {ti_le:.2f}%")
    if not ds_sai:
        print("  -> Khong co mau nao bi sai.")
    else:
        print("  -> Cac mau bi sai:")
        for s in ds_sai:
            print(f"     {s}")

def main():
    try:
        with open("hoa30.txt", "r") as f:
            lines = [line.strip() for line in f if line.strip()]
    except FileNotFoundError:
        print("Khong the mo tep hoa30.txt", file=sys.stderr)
        return

    n, d = map(int, lines[0].split())
    D = []
    for i in range(1, n + 1):
        parts = lines[i].split()
        features = [float(x) for x in parts[:d]]
        nhan = parts[d]
        D.append((i, features, nhan))

    print("=== KET QUA DANH GIA LOOCV TREN HOA30.TXT ===")
    danh_gia_loocv(D, "Euclid", kc_euclid)
    danh_gia_loocv(D, "Manhattan", kc_manhattan)

if __name__ == "__main__":
    main()